File: Languages/Python/Withdrawal.py
import Transaction
# Inheriting from Transaction class
class Withdrawal(Transaction): 
    def __init__(self, userAccountNumber, atmScreen, atmBankDatabase, atmKeypad, atmCashDispenser):
        super().__init__(userAccountNumber, atmScreen, atmBankDatabase)
        self.amount = 0
        self.keypad = atmKeypad
        self.cashDispenser = atmCashDispenser
        self.CANCELED = 6

    def execute(self):
        cashDispensed = False
        availableBalance = 0

        bankDatabase = self.getBankDatabase()
        screen = self.getScreen()

        while not cashDispensed:
            self.amount = self.display_menu_of_amounts()

            if self.amount != self.CANCELED:
                availableBalance = bankDatabase.get_available_balance(self.get_account_number())

                if self.amount <= availableBalance:
                    if self.cashDispenser.is_sufficient_cash_available(self.amount):
                        bankDatabase.debit(self.get_account_number(), self.amount)
                        self.cashDispenser.dispense_cash(self.amount)
                        cashDispensed = True
                        screen.display_message_line("\nYour cash has been dispensed. Please take your cash now.")
                    else:
                        screen.display_message_line("\nInsufficient cash available in the ATM." + "\n\nPlease choose a smaller amount.")
                else:
                    screen.display_message_line("\nInsufficient funds in your account." + "\n\nPlease choose a smaller amount.")
            else:
                screen.display_message_line("\nCanceling transaction...")
                return

    def display_menu_of_amounts(self):
        userChoice = 0
        screen = self.getScreen()

        amounts = [0, 20, 40, 60, 100, 200]

        while userChoice == 0:
            screen.display_message_line("\nWithdrawal Menu: ")
            screen.display_message_line("1 - $20")
            screen.display_message_line("2 - $40")
            screen.display_message_line("3 - $60")
            screen.display_message_line("4 - $100")
            screen.display_message_line("5 - $200")
            screen.display_message_line("6 - Cancel transaction")
            screen.display_message("\nChoose a withdrawal amount: ")

            input = self.keypad.get_input()

            if input in [1, 2, 3, 4, 5, 6]:
                userChoice = amounts[input]
            else:
                screen.display_message_line("\nInvalid selection. Try again.")

        return userChoice

File: Languages/Java/Withdrawal.java
package Languages.Java;

public class Withdrawal extends Transaction{ 
    private double amount; 
    private Keypad keypad; 
    private CashDispenser cashDispenser; 
    private final static int CANCELED = 6; 

    public Withdrawal(int userAccountNumber, Screen atmScreen, BankDatabase atmBankDatabase, Keypad atmKeypad, CashDispenser atmCashDispenser) {
        super(userAccountNumber, atmScreen, atmBankDatabase); 
        keypad = atmKeypad; 
        cashDispenser = atmCashDispenser; 
    }

    @Override 
    public void execute() {
        boolean cashDispensed = false; 
        double availableBalance; 


        //  get references to the bankdatabase and screen 
        BankDatabase bankDatabase = getBankDatabase(); 
        Screen screen = getScreen(); 


        // loop until cash is dispensed or user cancels 
        // AN ALGORITHM 
        do {
            // obtain a chosen withdrawal amount from the user 
            amount = displayMenuOfAmounts(); 

            // check whether user chose a withdrawal amount or canceled 
            if (amount != CANCELED) {
                // get available balance of account involved 
                availableBalance = bankDatabase.getAvailableBalance(getAccountNumber()); 

                // check whether user has enough money in the account 
                if (amount <= availableBalance) {
                    // check whether the cash dispenser has enough money 
                    if (cashDispenser.isSufficientCashAvailable(amount)) {
                        // update the account involved to reflect the withdrawal 
                        bankDatabase.debit(getAccountNumber(), amount); 

                        cashDispenser.dispenseCash(amount); // dispense cash 
                        cashDispensed = true; // cash was dispensed so loop ends 

                        // instruct user to take cash 
                        screen.displayMessageLine("\nYour cash has been " + "dispensed. Please take your cash now."); 
                    } else { // cash dispenser does not have enough cash 
                        screen.displayMessageLine("\nInsufficient cash available in the ATM." + "\n\nPlease choose a smaller amount."); 
                    }
                } else { // not enough money available in user's account 
                    screen.displayMessageLine("\nInsufficient funds in your account." + "\n\nPlease choose a smaller amount."); 
                }
            } else { // user chose cancel menu option 
                screen.displayMessageLine("\nCanceling transaction..."); 
                return; // return to main menu because user canceled 
            }
        } while (!cashDispensed);
        
    } 

    // method to display menu of amounts 
    private int displayMenuOfAmounts() {
        int userChoice = 0; 
        Screen screen = getScreen(); 

        int[] amounts = {0, 20, 40, 60, 100, 200}; 

        while (userChoice == 0) {
            screen.displayMessageLine("\nWithdrawal Menu: ");
            screen.displayMessageLine("1 - $20");
            screen.displayMessageLine("2 - $40");
            screen.displayMessageLine("3 - $60");
            screen.displayMessageLine("4 - $100");
            screen.displayMessageLine("5 - $200");
            screen.displayMessageLine("6 - Cancel transaction");
            screen.displayMessage("\nChoose a withdrawal amount: ");

            int input = keypad.getInput(); 

            switch (input) {
                case 1: 
                case 2: 
                case 3: 
                case 4: 
                case 5: 
                case 6: 
                    userChoice = amounts[input]; 
                    break; 
                default: 
                    screen.displayMessageLine("\nInvalid selection. Try again.");
            }
        }
        return userChoice; 
    }
    
}
